File: ui/src/server/networkPolicy.ts
import dns from 'dns/promises';
import net from 'net';
import { db } from './db';

export const OFFLINE_MODE_SETTING_KEY = 'OFFLINE_MODE';
export const OFFLINE_ALLOWED_HOSTS_ENV = 'AITK_OFFLINE_ALLOWED_HOSTS';

const OFFLINE_MODE_ENV_KEYS = ['AITK_OFFLINE_MODE', 'AI_TOOLKIT_OFFLINE_MODE'] as const;
const REMOTE_OLLAMA_WORKERS_SETTING_KEY = 'REMOTE_OLLAMA_WORKERS';
const TRUTHY_VALUES = new Set(['1', 'true', 'yes', 'on']);
const REDIRECT_STATUSES = new Set([301, 302, 303, 307, 308]);
const MAX_OFFLINE_REDIRECTS = 20;

const localPrivateBlockList = new net.BlockList();
localPrivateBlockList.addSubnet('127.0.0.0', 8, 'ipv4');
localPrivateBlockList.addSubnet('10.0.0.0', 8, 'ipv4');
localPrivateBlockList.addSubnet('172.16.0.0', 12, 'ipv4');
localPrivateBlockList.addSubnet('192.168.0.0', 16, 'ipv4');
localPrivateBlockList.addSubnet('169.254.0.0', 16, 'ipv4');
localPrivateBlockList.addAddress('::1', 'ipv6');
localPrivateBlockList.addSubnet('fc00::', 7, 'ipv6');
localPrivateBlockList.addSubnet('fe80::', 10, 'ipv6');

export class OfflineModeError extends Error {
  status = 403;
  code = 'OFFLINE_MODE_BLOCKED';

  constructor(message: string) {
    super(message);
    this.name = 'OfflineModeError';
  }
}

function boolValue(value: unknown, defaultValue = false) {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'string') {
    const normalized = value.trim().toLowerCase();
    if (['false', '0', 'off', 'disabled', 'no'].includes(normalized)) return false;
    if (['true', '1', 'on', 'enabled', 'yes'].includes(normalized)) return true;
  }
  return defaultValue;
}

export function isOfflineModeLockedByEnv(env: NodeJS.ProcessEnv = process.env) {
  return OFFLINE_MODE_ENV_KEYS.some(key =>
    TRUTHY_VALUES.has(
      String(env[key] || '')
        .trim()
        .toLowerCase(),
    ),
  );
}

export async function isOfflineModeEnabled() {
  if (isOfflineModeLockedByEnv()) return true;
  const row = await db.settings.get(OFFLINE_MODE_SETTING_KEY);
  return boolValue(row?.value, false);
}

export async function getOfflineModeState() {
  const locked = isOfflineModeLockedByEnv();
  const row = await db.settings.get(OFFLINE_MODE_SETTING_KEY);
  return {
    enabled: locked || boolValue(row?.value, false),
    lockedByEnv: locked,
  };
}

export function normalizeHostname(value: unknown) {
  const hostname = String(value || '')
    .trim()
    .toLowerCase()
    .replace(/^\[|\]$/g, '');
  return hostname.endsWith('.') ? hostname.slice(0, -1) : hostname;
}

function hostFromUrl(value: unknown) {
  try {
    return normalizeHostname(new URL(String(value)).hostname);
  } catch {
    return '';
  }
}

function isIpAddress(value: string) {
  return net.isIP(value) !== 0;
}

function ipv4MappedAddress(value: string) {
  const match = value.match(/^::ffff:(\d+\.\d+\.\d+\.\d+)$/i);
  return match?.[1] || null;
}

export function isLocalPrivateIp(value: string) {
  const normalized = normalizeHostname(value);
  if (!normalized) return false;

  const mapped = ipv4MappedAddress(normalized);
  const address = mapped || normalized;
  const family = net.isIP(address);
  if (family === 0) return false;
  if (address === '0.0.0.0' || address === '::') return false;
  return localPrivateBlockList.check(address, family === 4 ? 'ipv4' : 'ipv6');
}

async function resolveHostAddresses(hostname: string) {
  const host = normalizeHostname(hostname);
  if (!host) return [];
  if (isIpAddress(host)) return [host];
  const addresses = await dns.lookup(host, { all: true, verbatim: true });
  return Array.from(new Set(addresses.map(item => item.address)));
}

function envAllowedHostnames() {
  return new Set(
    String(process.env[OFFLINE_ALLOWED_HOSTS_ENV] || '')
      .split(',')
      .map(normalizeHostname)
      .filter(Boolean),
  );
}

type RawRemoteOllamaWorker = {
  base_url?: string;
  offline_bypass_enabled?: boolean;
};

export async function getOfflineBypassHostnames() {
  const hosts = envAllowedHostnames();

  const workerNodes = await db.workerNodes.list().catch(() => []);
  for (const worker of workerNodes) {
    if ((worker as any).offline_bypass_enabled) {
      const host = hostFromUrl(worker.base_url);
      if (host) hosts.add(host);
    }
  }

  const remoteOllamaRow = await db.settings.get(REMOTE_OLLAMA_WORKERS_SETTING_KEY).catch(() => null);
  if (remoteOllamaRow?.value) {
    try {
      const workers = JSON.parse(remoteOllamaRow.value) as RawRemoteOllamaWorker[];
      if (Array.isArray(workers)) {
        for (const worker of workers) {
          if (worker?.offline_bypass_enabled) {
            const host = hostFromUrl(worker.base_url);
            if (host) hosts.add(host);
          }
        }
      }
    } catch {
      // Ignore malformed legacy settings; they should not grant bypasses.
    }
  }

  return hosts;
}

function describeUrl(url: URL) {
  return `${url.protocol}//${url.host}`;
}

export async function assertUrlAllowedByOfflineMode(input: string | URL, feature = 'request') {
  if (!(await isOfflineModeEnabled())) return;

  const url = input instanceof URL ? input : new URL(String(input));
  if (!['http:', 'https:', 'ws:', 'wss:'].includes(url.protocol)) {
    throw new OfflineModeError(`Offline mode blocked ${feature}: unsupported URL protocol ${url.protocol}`);
  }

  const hostname = normalizeHostname(url.hostname);
  const bypassHosts = await getOfflineBypassHostnames();
  if (bypassHosts.has(hostname)) return;

  let addresses: string[];
  try {
    addresses = await resolveHostAddresses(hostname);
  } catch (error) {
    throw new OfflineModeError(
      `Offline mode blocked ${feature} to ${describeUrl(url)}: DNS lookup failed${
        error instanceof Error && error.message ? `: ${error.message}` : ''
      }`,
    );
  }

  if (!addresses.length) {
    throw new OfflineModeError(
      `Offline mode blocked ${feature} to ${describeUrl(url)}: DNS lookup returned no addresses`,
    );
  }

  const blocked = addresses.filter(address => !isLocalPrivateIp(address));
  if (blocked.length > 0) {
    throw new OfflineModeError(
      `Offline mode blocked ${feature} to ${describeUrl(url)}: resolved outside local/private IP space (${blocked.join(', ')})`,
    );
  }
}

function requestUrl(input: RequestInfo | URL) {
  if (typeof input === 'string' || input instanceof URL) return String(input);
  if (typeof Request !== 'undefined' && input instanceof Request) return input.url;
  return String(input);
}

function requestMethod(input: RequestInfo | URL, init?: RequestInit) {
  const method =
    init?.method || (typeof Request !== 'undefined' && input instanceof Request ? input.method : undefined) || 'GET';
  return method.toUpperCase();
}

function cloneHeadersWithoutBodyHeaders(headers: Headers) {
  const next = new Headers(headers);
  next.delete('content-length');
  next.delete('content-type');
  return next;
}

function redirectTarget(response: Response, currentUrl: string) {
  if (!REDIRECT_STATUSES.has(response.status)) return null;
  const location = response.headers.get('location');
  if (!location) return null;
  return new URL(location, currentUrl).toString();
}

export async function guardedFetch(input: RequestInfo | URL, init?: RequestInit, feature = 'request') {
  if (!(await isOfflineModeEnabled())) {
    return fetch(input, init);
  }

  const url =
    typeof input === 'string' || input instanceof URL
      ? input
      : typeof Request !== 'undefined' && input instanceof Request
        ? input.url
        : String(input);
  await assertUrlAllowedByOfflineMode(url, feature);

  if (init?.redirect === 'manual' || init?.redirect === 'error') {
    return fetch(input, init);
  }

  let currentInput = input;
  let currentInit = init;
  let currentUrl = requestUrl(currentInput);
  let currentMethod = requestMethod(currentInput, currentInit);

  for (let redirectCount = 0; redirectCount <= MAX_OFFLINE_REDIRECTS; redirectCount += 1) {
    const response = await fetch(currentInput, { ...currentInit, redirect: 'manual' });
    const nextUrl = redirectTarget(response, currentUrl);
    if (!nextUrl) return response;

    await assertUrlAllowedByOfflineMode(nextUrl, `${feature} redirect`);

    if (redirectCount === MAX_OFFLINE_REDIRECTS) {
      throw new OfflineModeError(`Offline mode blocked ${feature}: too many redirects`);
    }

    const nextHeaders = new Headers(currentInit?.headers);
    const shouldSwitchToGet =
      response.status === 303 || ((response.status === 301 || response.status === 302) && currentMethod === 'POST');
    if (shouldSwitchToGet) {
      currentInit = {
        ...currentInit,
        method: 'GET',
        body: undefined,
        headers: cloneHeadersWithoutBodyHeaders(nextHeaders),
      };
      currentMethod = 'GET';
    } else {
      currentInit = {
        ...currentInit,
        headers: nextHeaders,
      };
    }
    currentInput = nextUrl;
    currentUrl = nextUrl;
  }

  throw new OfflineModeError(`Offline mode blocked ${feature}: too many redirects`);
}

export function offlineChildProcessEnv(enabled: boolean, allowedHosts: Iterable<string> = []) {
  if (!enabled) return {};
  const hosts = Array.from(new Set(Array.from(allowedHosts).map(normalizeHostname).filter(Boolean)));
  return {
    AITK_OFFLINE_MODE: '1',
    HF_HUB_OFFLINE: '1',
    TRANSFORMERS_OFFLINE: '1',
    HF_DATASETS_OFFLINE: '1',
    ...(hosts.length > 0 ? { [OFFLINE_ALLOWED_HOSTS_ENV]: hosts.join(',') } : {}),
  };
}

export function hostnamesFromUrls(values: Iterable<string | null | undefined>) {
  const hosts = new Set<string>();
  for (const value of values) {
    const host = value ? hostFromUrl(value) : '';
    if (host) hosts.add(host);
  }
  return hosts;
}

File: toolkit/network_policy.py
import os
import socket
from ipaddress import ip_address, ip_network
from typing import Iterable
from urllib.parse import urlparse


OFFLINE_MODE_ENV_VARS = ("AITK_OFFLINE_MODE", "AI_TOOLKIT_OFFLINE_MODE")
OFFLINE_ALLOWED_HOSTS_ENV = "AITK_OFFLINE_ALLOWED_HOSTS"
HF_OFFLINE_ENV_VARS = ("HF_HUB_OFFLINE", "TRANSFORMERS_OFFLINE", "HF_DATASETS_OFFLINE")

_TRUTHY_VALUES = {"1", "ON", "TRUE", "YES"}
_LOCAL_PRIVATE_NETWORKS = tuple(
    ip_network(value)
    for value in (
        "127.0.0.0/8",
        "10.0.0.0/8",
        "172.16.0.0/12",
        "192.168.0.0/16",
        "169.254.0.0/16",
        "::1/128",
        "fc00::/7",
        "fe80::/10",
    )
)

_installed = False
_allowed_resolved_addresses: set[str] = set()
_original_getaddrinfo = socket.getaddrinfo
_original_create_connection = socket.create_connection
_original_socket_connect = socket.socket.connect


class OfflineNetworkError(RuntimeError):
    pass


def env_flag(name: str) -> bool:
    value = os.getenv(name)
    return value is not None and value.strip().upper() in _TRUTHY_VALUES


def is_offline_mode_enabled() -> bool:
    return any(env_flag(name) for name in OFFLINE_MODE_ENV_VARS)


def configure_offline_environment() -> bool:
    if not is_offline_mode_enabled():
        return False

    for name in HF_OFFLINE_ENV_VARS:
        os.environ[name] = os.environ.get(name) or "1"
    os.environ["DISABLE_TELEMETRY"] = os.environ.get("DISABLE_TELEMETRY") or "YES"
    return True


def _normalize_hostname(value: object) -> str:
    hostname = str(value or "").strip().lower().strip("[]")
    return hostname[:-1] if hostname.endswith(".") else hostname


def _allowed_hosts() -> set[str]:
    raw = os.getenv(OFFLINE_ALLOWED_HOSTS_ENV, "")
    return {_normalize_hostname(item) for item in raw.split(",") if _normalize_hostname(item)}


def _is_allowed_host(hostname: str) -> bool:
    return _normalize_hostname(hostname) in _allowed_hosts()


def _record_allowed_addresses(addresses: Iterable[str]):
    for address in addresses:
        normalized = str(address or "").strip()
        if normalized:
            _allowed_resolved_addresses.add(normalized)


def is_local_private_ip(value: object) -> bool:
    try:
        parsed = ip_address(str(value).strip())
    except ValueError:
        return False

    if parsed.is_unspecified or parsed.is_multicast:
        return False

    if parsed.version == 6 and parsed.ipv4_mapped is not None:
        parsed = parsed.ipv4_mapped

    return any(parsed in network for network in _LOCAL_PRIVATE_NETWORKS)


def is_allowed_ip(value: object) -> bool:
    address = str(value or "").strip()
    return address in _allowed_resolved_addresses or is_local_private_ip(address)


def _resolve_addresses(hostname: str, port: int | None = None) -> list[str]:
    host = _normalize_hostname(hostname)
    if not host:
        return []

    try:
        ip_address(host)
        return [host]
    except ValueError:
        pass

    infos = _original_getaddrinfo(host, port, type=socket.SOCK_STREAM)
    addresses: list[str] = []
    for info in infos:
        sockaddr = info[4]
        if sockaddr:
            addresses.append(str(sockaddr[0]))
    return list(dict.fromkeys(addresses))


def assert_host_allowed(hostname: str, port: int | None = None, feature: str = "request") -> None:
    if not is_offline_mode_enabled():
        return

    host = _normalize_hostname(hostname)
    if not host:
        raise OfflineNetworkError(f"Offline mode blocked {feature}: missing hostname")

    if _is_allowed_host(host):
        addresses = _resolve_addresses(host, port)
        _record_allowed_addresses(addresses)
        return

    try:
        addresses = _resolve_addresses(host, port)
    except OSError as exc:
        raise OfflineNetworkError(f"Offline mode blocked {feature} to {host}: DNS lookup failed: {exc}") from exc

    if not addresses:
        raise OfflineNetworkError(f"Offline mode blocked {feature} to {host}: DNS lookup returned no addresses")

    blocked = [address for address in addresses if not is_allowed_ip(address)]
    if blocked:
        raise OfflineNetworkError(
            f"Offline mode blocked {feature} to {host}: resolved outside local/private IP space ({', '.join(blocked)})"
        )


def assert_url_allowed(url: str, feature: str = "request") -> None:
    if not is_offline_mode_enabled():
        return

    parsed = urlparse(str(url))
    if parsed.scheme and parsed.scheme.lower() not in {"http", "https", "ws", "wss"}:
        raise OfflineNetworkError(f"Offline mode blocked {feature}: unsupported URL protocol {parsed.scheme}")

    assert_host_allowed(parsed.hostname or "", parsed.port, feature)


def _guarded_getaddrinfo(host, port, *args, **kwargs):
    if is_offline_mode_enabled() and host is not None:
        assert_host_allowed(str(host), int(port) if port is not None else None, "DNS lookup")
    return _original_getaddrinfo(host, port, *args, **kwargs)


def _guarded_socket_connect(self, address):
    if is_offline_mode_enabled() and isinstance(address, tuple) and len(address) >= 2:
        host = str(address[0])
        port = int(address[1])
        if is_allowed_ip(host):
            return _original_socket_connect(self, address)
        assert_host_allowed(host, port, "socket connection")
    return _original_socket_connect(self, address)


def _guarded_create_connection(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, source_address=None, *args, **kwargs):
    if is_offline_mode_enabled() and isinstance(address, tuple) and len(address) >= 2:
        assert_host_allowed(str(address[0]), int(address[1]), "socket connection")
    return _original_create_connection(address, timeout, source_address, *args, **kwargs)


def install_offline_network_guard() -> bool:
    global _installed

    configure_offline_environment()
    if _installed or not is_offline_mode_enabled():
        return _installed

    socket.getaddrinfo = _guarded_getaddrinfo
    socket.socket.connect = _guarded_socket_connect
    socket.create_connection = _guarded_create_connection
    _installed = True
    return True
